// @minka/ledger-sdk/types/src/lib/records/ledger-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventSignals = exports.EventSignal = void 0;
/**
 * Name of the signal of the event.
 */
var EventSignal;
(function (EventSignal) {
    // --- General
    EventSignal["BalanceReceived"] = "balance-received";
    EventSignal["WalletLimited"] = "wallet-limited";
    // --- Create
    EventSignal["EffectCreated"] = "effect-created";
    EventSignal["LedgerCreated"] = "ledger-created";
    EventSignal["IntentCreated"] = "intent-created";
    EventSignal["SignerCreated"] = "signer-created";
    EventSignal["SymbolCreated"] = "symbol-created";
    EventSignal["PolicyCreated"] = "policy-created";
    EventSignal["SchemaCreated"] = "schema-created";
    EventSignal["WalletCreated"] = "wallet-created";
    EventSignal["BridgeCreated"] = "bridge-created";
    EventSignal["CircleCreated"] = "circle-created";
    EventSignal["RequestCreated"] = "request-created";
    EventSignal["AnchorCreated"] = "anchor-created";
    EventSignal["BridgeEntryCreated"] = "bridge-entry-created";
    // --- Update
    EventSignal["EffectUpdated"] = "effect-updated";
    EventSignal["LedgerUpdated"] = "ledger-updated";
    EventSignal["IntentUpdated"] = "intent-updated";
    EventSignal["SignerUpdated"] = "signer-updated";
    EventSignal["SymbolUpdated"] = "symbol-updated";
    EventSignal["PolicyUpdated"] = "policy-updated";
    EventSignal["SchemaUpdated"] = "schema-updated";
    EventSignal["WalletUpdated"] = "wallet-updated";
    EventSignal["BridgeUpdated"] = "bridge-updated";
    EventSignal["CircleUpdated"] = "circle-updated";
    EventSignal["RequestUpdated"] = "request-updated";
    EventSignal["BridgeEntryUpdated"] = "bridge-entry-updated";
    EventSignal["AnchorUpdated"] = "anchor-updated";
    // --- Drop
    EventSignal["AnchorDropped"] = "anchor-dropped";
})(EventSignal = exports.EventSignal || (exports.EventSignal = {}));
/**
 * All possible event signals.
 */
exports.EventSignals = Object.values(EventSignal);
