// @minka/ledger-sdk/types/src/lib/records/ledger-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaFormats = exports.SchemaFormat = exports.SchemaRecords = exports.SchemaRecord = void 0;
/**
 * Records that can be validated by using
 * the schema.
 */
var SchemaRecord;
(function (SchemaRecord) {
    SchemaRecord["Symbol"] = "symbol";
    SchemaRecord["Signer"] = "signer";
    SchemaRecord["Wallet"] = "wallet";
    SchemaRecord["Intent"] = "intent";
    SchemaRecord["Effect"] = "effect";
    SchemaRecord["Bridge"] = "bridge";
    SchemaRecord["Circle"] = "circle";
    SchemaRecord["Policy"] = "policy";
    SchemaRecord["Anchor"] = "anchor";
    SchemaRecord["AnchorLookup"] = "anchor-lookup";
})(SchemaRecord = exports.SchemaRecord || (exports.SchemaRecord = {}));
/**
 * Schema Record values
 */
exports.SchemaRecords = Object.values(SchemaRecord);
/**
 * Available schema formats.
 */
var SchemaFormat;
(function (SchemaFormat) {
    SchemaFormat["JsonSchema"] = "json-schema";
})(SchemaFormat = exports.SchemaFormat || (exports.SchemaFormat = {}));
/**
 * Schema Format values
 */
exports.SchemaFormats = Object.values(SchemaFormat);
